src/k2/module/k23si/Module.cpp

```diff
@@ -159,24 +163,192 @@ _makeReadOK(dto::DataRecord* rec) {
+        while (key_it != _indexer.end() && key_it->first > request.key) {
+            scanAdvance(key_it, _indexer, request.reverseDirection);
+        }
+    }
+
+    if (key_it != _indexer.end() && key_it->first.schemaName != request.key.schemaName) {
+        key_it = _indexer.end();
+    }
+
+    for (; key_it != _indexer.end(); scanAdvance(key_it, _indexer, request.reverseDirection)) {
+        // Extra Termination conditions
+        if (!request.reverseDirection && key_it->first >= request.endKey &&
+                   request.endKey.partitionKey != "") {
+            break;
+        } else if (request.reverseDirection && key_it->first <= request.endKey) {
+            break;
+        } else if (request.recordLimit >= 0 && response.results.size() == (uint32_t)request.recordLimit) {
+            break;
+        } else if (response.results.size() == _config.paginationLimit()) {
+            break;
+        }
+
+
+        auto& versions = key_it->second;
+        auto viter = versions.begin();
+        // position the version iterator at the version we should be returning
+        while(viter != versions.end() && request.mtr.timestamp.compareCertain(viter->txnId.mtr.timestamp) < 0) {
+            ++viter;
+        }
+        if (viter == versions.end()) {
+            // happy case: we either had no versions, or all versions were newer than the requested timestamp
+            continue;
+        }
+
+        // happy case: either committed, or txn is reading its own write
+        if ((viter->status == dto::DataRecord::Committed || viter->txnId.mtr == request.mtr) 
+                    && !viter->isTombstone) {
+            // TODO apply filter and projection
+            response.results.push_back(viter->value.share());
+            continue;
+        } else if (viter->status == dto::DataRecord::Committed || viter->txnId.mtr == request.mtr) {
+            continue;
+        }
+
+        // If we get here it is a conflict, first decide to push or return early
+        if (response.results.size() >= _config.queryPushLimit()) {
+            break;
+        }
+
+        // Do a push but we need to save our place in the query
+        // TODO we can test the filter condition against the WI and last committed version and possibly
+        // avoid a push
+        // Must update read cache before doing an async operation
+        request.reverseDirection ? 
+            _readCache->insertInterval(key_it->first, request.key, request.mtr.timestamp) :
+            _readCache->insertInterval(request.key, key_it->first, request.mtr.timestamp);
+
+        K2DEBUG("About to PUSH in query request");
+        auto sitMTR = viter->txnId.mtr;
+        return _doPush(request.collectionName, viter->txnId, request.mtr, deadline)
+        .then([this, curKey=key_it->first, sitMTR, request=std::move(request), 
+                        resp=std::move(response), deadline](auto&& winnerMTR) mutable {
+            if (winnerMTR == sitMTR) {
```

Comment:
I wonder if we could refactor this logic in the _doPush method - we always want to cleanup the loser WI i think

Reply:
I added a new task for this.